### PR TITLE
Implement CTLineCreateWithAttributedString, CTRunGetGlyphs, CTRun Get…

### PR DESCRIPTION
**Pull request by aballway**
…Ptr functions, and add CTLine tests

Has some temporary workarounds until we get CTFont and glyphOrigins working properly, and some code will have to change when we change the structures we use.

### CODE REVIEW

File: Frameworks/CoreText/CTLine.mm [hunk 1]
```diff
@@ -38,7 +44,7 @@ - (instancetype)copyWithZone:(NSZone*)zone {
     ret->_ascent = _ascent;
     ret->_descent = _descent;
     ret->_leading = _leading;
-    ret->_runs = [_runs copy];
```
Comment (mnithish) [inline Frameworks/CoreText/CTLine.mm]: >ret->_runs = [_runs copy]; [](start = 4, length = 26)

Why wouldn't copy work here?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 2]
```diff
@@ -327,6 +415,65 @@ static void _GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameSi
 }
 
 /**
+ * Helper method to create _CTLine object given a CFAttributedStringRef
+ *
+ * @parameter string CFAttributedStringRef containing text and styling
+ *
+ * @return Unmutable array of _CTLine objects created with the requested parameters.
+ */
+static _CTLine* _DWriteGetLine(CFAttributedStringRef string) {
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: >_DWriteGetLine [](start = 16, length = 14)

Can we avoid code duplication between _DWriteGetLine and _DWriteGetLines? We can create a _CTTypesetter object with the given CFAttributedStringRef and we should be able to call DWriteGetLines directly from here

**Comment by mnithish** [comment]
>     CGFloat _width;

Looks like we need this field. Move it above this comment

---
Refers to: Frameworks/include/CoreTextInternal.h:68 in da34602. [](commit_id = da346025e07d57a07dc31e3e5170bed12d78f78e, deletion_comment = False)

File: Frameworks/CoreText/CTLine.mm [hunk 3]
```diff
@@ -248,12 +242,8 @@ void CTLineDraw(CTLineRef lineRef, CGContextRef ctx) {
  @Status Interoperable
 */
 CFIndex CTLineGetGlyphCount(CTLineRef lineRef) {
-    if (lineRef == nil) {
-        return 0;
-    }
-
     _CTLine* line = (_CTLine*)lineRef;
-    return (line->_strRange).length;
+    return line ? (line->_strRange).length : 0;
```
Comment (mnithish) [inline Frameworks/CoreText/CTLine.mm]: >(line->_strRange).length [](start = 18, length = 24)

I just learnt that character count != glyph count for all languages. We should get the total glyph count from the glyph runs that actually make the line. May be it is worth caching this as a new member in _CTLine

File: Frameworks/CoreText/CTRun.mm [hunk 4]
```diff
@@ -79,101 +88,122 @@ CTRunStatus CTRunGetStatus(CTRunRef run) {
...
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 const CGGlyph* CTRunGetGlyphsPtr(CTRunRef run) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    _CTRun* runPtr = static_cast<_CTRun*>(run);
+    return runPtr && runPtr->_glyphs.size() ? runPtr->_glyphs.data() : NULL;
```
Comment (mnithish) [inline Frameworks/CoreText/CTRun.mm]: >NULL [](start = 71, length = 4)

nit: nullptr

File: Frameworks/CoreText/CTRun.mm [hunk 5]
```diff
@@ -79,101 +88,122 @@ CTRunStatus CTRunGetStatus(CTRunRef run) {
...
+    return runPtr && runPtr->_glyphOrigins.size() ? runPtr->_glyphOrigins.data() : NULL;
 }
 
 /**
- @Status Caveat
- @Notes Only the starting position of the glyph run is handled.
+ @Status Interoperable
+ @Notes
 */
 void CTRunGetPositions(CTRunRef run, CFRange runRange, CGPoint* outPositions) {
```
Comment (mnithish) [inline Frameworks/CoreText/CTRun.mm]: >outPositions [](start = 64, length = 12)

Check for null

Comment (mnithish) [inline Frameworks/CoreText/CTRun.mm]: I guess you are OK here. There is no return status we can provide so to blow up here is fine

---
In reply to: [78227492](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78227492) [](ancestors = 78227492)

File: tests/unittests/CoreText/CTTypesetterTests.mm [hunk 6]
```diff
@@ -45,15 +45,20 @@
     EXPECT_EQ_MSG(13,
                   index,
                   "SuggestLineBreak should end line at space when soft line break is possible and the text is longer than the width");
+
+    CFRelease(ts);
 }
 
+/*
```
Comment (rajsesh-msft) [inline tests/unittests/CoreText/CTTypesetterTests.mm]: Can you add a TODO with issue number which will enable these tests again?

File: tests/unittests/CoreText/CTLineTests.mm [hunk 7]
```diff
@@ -98,4 +98,141 @@
...
+
+    EXPECT_NE(runs, nil);
+    EXPECT_EQ_MSG(2, CFArrayGetCount(runs), "CTLine should create multiple runs when attributed string has multiple attribute ranges");
+    EXPECT_EQ(3, CTRunGetGlyphCount(static_cast<CTRunRef>(CFArrayGetValueAtIndex(runs, 0))));
+    EXPECT_EQ(2, CTRunGetGlyphCount(static_cast<CTRunRef>(CFArrayGetValueAtIndex(runs, 1))));
+
+    // Glyphs for h, e, and l respectively
+    const CGGlyph* helGlyphs = CTRunGetGlyphsPtr(static_cast<CTRunRef>(CFArrayGetValueAtIndex(runs, 0)));
+    ASSERT_NE(nullptr, helGlyphs);
+    EXPECT_EQ(75, *helGlyphs);
```
Comment (mnithish) [inline tests/unittests/CoreText/CTLineTests.mm]: >EXPECT_EQ(75, *helGlyphs); [](start = 4, length = 26)

This test will fail if we change our default font or the locale. Might be a good idea to set the font name and locale for this test for such hard coded values

**Comment by mnithish** [comment]
:clock1:

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 8]
```diff
@@ -369,6 +518,26 @@ static void _GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameSi
...
+            }
+            run->_glyphs.resize(glyphRunDetails._glyphRuns[i]._glyphIndices.size());
+            std::vector<WORD> characters(run->_range.length);
+            [run->_stringFragment getCharacters:characters.data()];
+            CTFontGetGlyphsForCharacters(static_cast<CTFontRef>(font), characters.data(), run->_glyphs.data(), run->_range.length);
+
+            // TODO::
+            // This is a temp workaround until we can have actual glyph origins
+            for (auto i : run->_glyphAdvances) {
+                run->_glyphOrigins.emplace_back(CGPoint{ line->_width + i.width, 0 });
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: Should the first glyph origin be run->_xPos + ?

File: Frameworks/CoreText/CTFont.mm [hunk 9]
```diff
@@ -621,12 +621,17 @@ CFArrayRef CTFontCopyFeatureSettings(CTFontRef font) {
...
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 bool CTFontGetGlyphsForCharacters(CTFontRef font, const UniChar characters[], CGGlyph glyphs[], CFIndex count) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    bool succeeded = true;
+    FT_Face face = (FT_Face)[(UIFont*)font _sizingFontHandle];
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTFont.mm]: ? Why are we adding more freetype here?

File: Frameworks/CoreText/CTRun.mm [hunk 10]
```diff
@@ -44,18 +44,27 @@ - (id)init {
     return self;
 }
 
+- (instancetype)copyWithZone:(NSZone*)zone {
+    _CTRun* ret = [_CTRun new];
+    ret->_attributes = [_attributes copy];
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTRun.mm]: >_attributes = [_attributes copy]; [](start = 9, length = 33)

I don't think CoreText has ARC enabled does it? If not, this and the copy below would leak. Use .attach() instead of =, or call autorelease

File: Frameworks/CoreText/CTRun.mm [hunk 11]
```diff
@@ -79,101 +88,122 @@ CTRunStatus CTRunGetStatus(CTRunRef run) {
...
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 void CTRunGetGlyphs(CTRunRef run, CFRange range, CGGlyph buffer[]) {
-    UNIMPLEMENTED();
+    _CTRun* runPtr = static_cast<_CTRun*>(run);
+    if (runPtr) {
+        if (range.location < 0) {
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTRun.mm]: Are these just bounding the range? (If so, I find that it's more elegant to do so using std::max/min)

Comment (ms-jihua) [inline Frameworks/CoreText/CTRun.mm]: You're actually repeating this pattern a lot in this file - make a helper for it?

---
In reply to: [78249297](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78249297) [](ancestors = 78249297)

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm [hunk 12]
```diff
@@ -28,8 +28,6 @@ @implementation CTTruncatedLineTestView {
 }
 
 - (void)drawRect:(CGRect)rect {
-    UIColor* color = [UIColor blueColor];
```
Comment (ms-jihua) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm]: didn't you already make this deletion in another commit? I could've sworn I saw this already...

File: Frameworks/CoreText/CTFont.mm (hunk 9 above)
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFont.mm]: We have to only use DWrite apis here - freetype will be removed after we migrate to DWrite.

---
In reply to: [78248109](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78248109) [](ancestors = 78248109)

File: Frameworks/CoreText/CTFrame.mm [hunk 13]
```diff
@@ -55,7 +55,7 @@ CFRange CTFrameGetVisibleStringRange(CTFrameRef frame) {
     _CTFrame* framePtr = static_cast<_CTFrame*>(frame);
     CFIndex count = 0;
     if (framePtr) {
-        for (CFIndex i = 0; i < framePtr->_lineOrigins.size() && framePtr->_lineOrigins[i].y > 0; ++i) {
+        for (CFIndex i = 0; i < [framePtr->_lines count] && static_cast<_CTLine*>(framePtr->_lines[i])->_lineOrigin.y >= 0; ++i) {
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFrame.mm]: >[framePtr->_lines count] && static_cast<_CTLine*>(framePtr->_lines[i])->_lineOrigin.y >= 0 [](start = 32, length = 90)

why the y >= 0? Also, it might make sense to split this inside the for loop so you can get rid of the objectAtIndex call below.

File: Frameworks/CoreText/CTLine.mm [hunk 14]
```diff
@@ -19,8 +19,14 @@
 #import "NSStringInternal.h"
 #import "CoreTextInternal.h"
 #import "CGContextInternal.h"
+#import "DWriteWrapper.h"
 #import <CoreText/CTTypesetter.h>
 
+#include <algorithm>
+
+static IWLazyClassLookup _LazyUIFont("UIFont");
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTLine.mm]: >_LazyUIFont [](start = 25, length = 11)

why is this here?

File: Frameworks/CoreText/CTLine.mm (hunk 1 above)
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTLine.mm]: if you need a mutable array, use mutableCopy

---
In reply to: [78225046](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78225046) [](ancestors = 78225046)

Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: copy by default doesn't copy the items, so each run would be sharing a reference to a string, which we would lose should the original run be dealloced

---
In reply to: [78260653](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78260653) [](ancestors = 78260653,78225046)

File: Frameworks/CoreText/CTRun.mm (hunk 4 above)
Comment (aballway) [inline Frameworks/CoreText/CTRun.mm]: the docs have these specified as NULL.  is there any case where a developer could be expecting NULL and nullptr would not work?

---
In reply to: [78227167](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78227167) [](ancestors = 78227167)

File: Frameworks/CoreText/CTFont.mm (hunk 9 above)
Comment (aballway) [inline Frameworks/CoreText/CTFont.mm]: This was just so I would be able to test my changes until the CTFont migration gets merged in.  I expect these to be overwritten by your changes

---
In reply to: [78260160](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78260160) [](ancestors = 78260160,78248109)

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm (hunk 12 above)
Comment (aballway) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCLineTestViewController.mm]: it may have been another file? I'm not sure

---
In reply to: [78249523](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78249523) [](ancestors = 78249523)

File: Frameworks/CoreText/CTLine.mm (hunk 1 above)
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: `-[NSArray copy]` absolutely returns a new array with _retained_ references to the items inside it. So long as either array is around, the objects inside them will live on.

File: Frameworks/CoreText/CTFrame.mm (hunk 13 above)
Comment (aballway) [inline Frameworks/CoreText/CTFrame.mm]: 0 would be the bottom of the frame, so any line with an origin >= 0 would be completely within the frame

---
In reply to: [78260477](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78260477) [](ancestors = 78260477)

File: Frameworks/CoreText/CTLine.mm [hunk 15]
```diff
@@ -19,8 +19,11 @@
 #import "NSStringInternal.h"
 #import "CoreTextInternal.h"
 #import "CGContextInternal.h"
+#import "DWriteWrapper.h"
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: reminder to both of us that I moved this to a different dir, so one of us will have to update this import depending on which of our changes goes in first

File: tests/unittests/CoreText/CTLineTests.mm [hunk 16]
```diff
@@ -98,4 +98,153 @@
...
+    // Glyphs for h, e, and l respectively
+    const CGGlyph* helGlyphs = CTRunGetGlyphsPtr(static_cast<CTRunRef>(CFArrayGetValueAtIndex(runs, 0)));
+    ASSERT_NE(nullptr, helGlyphs);
+
+    std::vector<WORD> characters(4);
+    [@"hel" getCharacters:characters.data()];
+    std::vector<CGGlyph> actualGlyphs(4);
+    UIFontDescriptor* fontDescriptor = [UIFontDescriptor fontDescriptorWithName:@"Times New Roman" size:40];
+    UIFont* font = [UIFont fontWithDescriptor:fontDescriptor size:40];
+    CTFontGetGlyphsForCharacters(static_cast<CTFontRef>(font), characters.data(), actualGlyphs.data(), 3);
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTLineTests.mm]: >3 [](start = 103, length = 1)

What happens if you feed a bad value here? (exceed range or negative length)

File: tests/unittests/CoreText/CTTypesetterTests.mm [hunk 17]
```diff
@@ -45,15 +45,20 @@
     EXPECT_EQ_MSG(13,
                   index,
                   "SuggestLineBreak should end line at space when soft line break is possible and the text is longer than the width");
+
+    CFRelease(ts);
 }
 
+/* TODO:: 998
 TEST(CTTypeSetter, CreateLineShouldTrimWhitespace) {
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTTypesetterTests.mm]: DISABLED_TEST works just as well - it's not as if these tests are failing to build, after all.

File: tests/unittests/CoreText/CTLineTests.mm [hunk 18]
```diff
@@ -98,4 +98,153 @@
...
+    EXPECT_EQ(0, CTLineGetGlyphCount(line));
+
+    runs = CTLineGetGlyphRuns(line);
+    EXPECT_NE(runs, nil);
+    EXPECT_EQ(0, CFArrayGetCount(runs));
+    CFRelease(runs);
+    CFRelease(line);
+}
+
+TEST(CTLine, CreateTruncatedLineTest) {
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTLineTests.mm]: nit: this test is a bit long/repetitive/looks possibly difficult to maintain? consider writing a helper for your validation and/or making this a parameterized test or breaking this down into multiple test cases.

File: Frameworks/CoreText/CTLine.mm (hunk 15 above)
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: or just make the same move yourself to make the merge easier on both of us :P

---
In reply to: [78466235](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78466235) [](ancestors = 78466235)

File: tests/unittests/CoreText/CTLineTests.mm (hunk 16 above)
Comment (aballway) [inline tests/unittests/CoreText/CTLineTests.mm]: For which method?  CTFont... doesn't use the __bound method, and I'd consider that outside of the scope of this test.

---
In reply to: [78467363](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78467363) [](ancestors = 78467363)

File: Frameworks/CoreText/CTLine.mm (hunk 15 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: I'd rather just change the line whenever it's merged than move all of them now.

---
In reply to: [78468576](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78468576) [](ancestors = 78468576,78466235)

File: tests/unittests/CoreText/CTLineTests.mm (hunk 16 above)
Comment (ms-jihua) [inline tests/unittests/CoreText/CTLineTests.mm]: nvm, should've read the function documentation first ^^;;;

---
In reply to: [78471645](https://github.com/Microsoft/WinObjC/pull/989#discussion_r78471645) [](ancestors = 78471645,78467363)

**Comment by ms-jihua** [comment]
:shipit:

**Comment by aballway** [comment]
Rebased on top of changes to font and drawing changes, and have more updates to tests.

File: build/CoreText/lib/CoreTextLib.vcxproj [hunk 19]
```diff
@@ -60,6 +60,10 @@
     <ClangCompile>
       <IncludePaths>$(StarboardBasePath)\Frameworks\include;$(StarboardBasePath)\include\xplat;$(StarboardBasePath)\deps\prebuilt\include;$(StarboardBasePath)\deps\prebuilt\include\freetype;</IncludePaths>
       <AdditionalOptions>"-DCORETEXT_IMPEXP= " %(AdditionalOptions)</AdditionalOptions>
+      <OtherCFlags>
+      </OtherCFlags>
+      <OtherCPlusPlusFlags>
+      </OtherCPlusPlusFlags>
```
Comment (mnithish) [inline build/CoreText/lib/CoreTextLib.vcxproj]: ??

Comment (aballway) [inline build/CoreText/lib/CoreTextLib.vcxproj]: I have no clue how that got happened

---
In reply to: [79678864](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79678864) [](ancestors = 79678864)

File: Frameworks/CoreText/CTFont.mm [hunk 20]
```diff
@@ -664,12 +664,20 @@ CFArrayRef CTFontCopyFeatureSettings(CTFontRef font) {
...
+    if (!fontRef || !characters || !glyphs || count <= 0) {
+        return false;
+    }
+
+    struct __CTFont* font = const_cast<struct __CTFont*>(fontRef);
+    std::vector<uint32_t> chars(characters, characters + count);
+    font->_dwriteFontFace->GetGlyphIndices(chars.data(), count, glyphs);
+
+    // Failed to get all glyphs if any of the indices are 0
+    return std::none_of(glyphs, glyphs + count, [](CGGlyph glyph) { return glyph == 0; });
```
Comment (mnithish) [inline Frameworks/CoreText/CTFont.mm]: > return std::none_of(glyphs, glyphs + count, [](CGGlyph glyph) { return glyph == 0; }); [](start = 3, length = 87)

Let us leave these to DirectWrite to tell us if there was an error or not.

File: Frameworks/CoreText/CTFramesetter.mm [hunk 21]
```diff
@@ -23,30 +23,31 @@
...
 }
 @end
 
-static _CTFrame* _createFrame(_CTFrameSetter* frameSetter, CGRect frameSize, CGSize* sizeOut, bool createFrame) {
+static _CTFrame* _createFrame(_CTFramesetter* framesetter, CGRect frameSize, CGSize* sizeOut, bool createFrame) {
     _CTFrame* ret = nil;
 
     if (createFrame) {
         ret = [_CTFrame new];
-        ret->_frameSetter = [frameSetter retain];
```
Comment (mnithish) [inline Frameworks/CoreText/CTFramesetter.mm]: >[frameSetter retain]; [](start = 28, length = 21)

Why have you removed the retain here? framesetter object is being handed over to CTFrame and it should hold a reference to it right?

File: Frameworks/CoreText/CTFrame.mm [hunk 22]
```diff
@@ -55,9 +56,10 @@ CFRange CTFrameGetVisibleStringRange(CTFrameRef frame) {
     _CTFrame* framePtr = static_cast<_CTFrame*>(frame);
     CFIndex count = 0;
     if (framePtr) {
-        for (CFIndex i = 0; i < framePtr->_lineOrigins.size() && framePtr->_lineOrigins[i].y > 0; ++i) {
-            // Lines are ordered vertically so we can stop once we find one outside of the frame
-            count += static_cast<_CTLine*>([framePtr->_lines objectAtIndex:i])->_strRange.length;
+        for (_CTLine* line in [framePtr->_lines objectEnumerator]) {
```
Comment (mnithish) [inline Frameworks/CoreText/CTFrame.mm]: >[framePtr->_lines objectEnumerator] [](start = 30, length = 35)

nit: framePtr->_lines is sufficient here

File: Frameworks/include/CoreTextInternal.h [hunk 23]
```diff
@@ -24,6 +24,20 @@
 #import <wrl/client.h>
 #include <COMIncludes_End.h>
 #import <vector>
+#import <algorithm>
+
+// Helper method for validating range and copying internal data to given outData
+template <typename T>
+void _boundedCopy(CFRange range, CFIndex size, const T inData[], T outData[]) {
```
Comment (mnithish) [inline Frameworks/include/CoreTextInternal.h]: >inData [](start = 55, length = 6)

Would be nice to check if inData is valid

File: Frameworks/CoreText/CTFramesetter.mm [hunk 24]
```diff
@@ -87,18 +88,22 @@ CTFrameRef CTFramesetterCreateFrame(CTFramesetterRef framesetter, CFRange string
     _CGPathGetBoundingBoxInternal(path, &frameSize);
 
     CGSize sizeOut;
-    _CTFrame* ret = _createFrame((_CTFrameSetter*)framesetter, frameSize, &sizeOut, true);
+    _CTFrame* ret = _createFrame((_CTFramesetter*)framesetter, frameSize, &sizeOut, true);
+    if (ret->_lineOrigins.size() == 0) {
+        for (_CTLine* line in [ret->_lines objectEnumerator]) {
+            ret->_lineOrigins.emplace_back(line->_lineOrigin);
+        }
+    }
```
Comment (mnithish) [inline Frameworks/CoreText/CTFramesetter.mm]: Let us move this code to _createFrame. Ideally when a CTFrame object is created it should have all its members setup properly.

**Comment by mnithish** [comment]
>         line->_runs = runs;

With your new change, we are leaking here. You did an attach of a mutable array in the CTLine INIT

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:454 in 8c41446. [](commit_id = 8c414467615be15f6033f0d28e81ac563d40ec1d, deletion_comment = False)

File: Frameworks/CoreText/CTFramesetter.mm (hunk 21 above)
Comment (aballway) [inline Frameworks/CoreText/CTFramesetter.mm]: I changed _framesetter to a StrongId<>

---
In reply to: [79681629](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79681629) [](ancestors = 79681629)

File: Frameworks/CoreText/CTLine.mm [hunk 25]
```diff
@@ -48,16 +60,7 @@ - (instancetype)copyWithZone:(NSZone*)zone {
...
-    NSString* str = [(NSAttributedString*)string string];
-    CFRange lineRange;
-    lineRange.location = 0;
-    lineRange.length = [str length];
-
-    _CTLine* line = [_CTLine new];
-    line->_strRange = lineRange;
-
-    return (CTLineRef)line;
+    return static_cast<CTLineRef>(_DWriteGetLine(string));
```
Comment (mnithish) [inline Frameworks/CoreText/CTLine.mm]: >_DWriteGetLine [](start = 34, length = 14)

The API says it we should return NULL on a failure but I see _DWriteGetLine is returning a new CTLine object

File: Frameworks/CoreText/CTFramesetter.mm (hunk 21 above)
Comment (mnithish) [inline Frameworks/CoreText/CTFramesetter.mm]: Ah I missed that

---
In reply to: [79685155](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79685155) [](ancestors = 79685155,79681629)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 26]
```diff
@@ -343,6 +357,24 @@ static void __GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameS
...
+ */
+static _CTLine* _DWriteGetLine(CFAttributedStringRef string) {
+    _CTTypesetter* typesetter = static_cast<_CTTypesetter*>(CTTypesetterCreateWithAttributedString(string));
+    CFRange range = CFRangeMake(0, CFAttributedStringGetLength(string));
+    NSArray<_CTLine*>* lines = _DWriteGetLines(typesetter, range, CGRectMake(0, 0, FLT_MAX, FLT_MAX));
+    if (lines.count > 0) {
+        return [lines[0] retain];
+    }
+
+    return [_CTLine new];
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: >[_CTLine new]; [](start = 11, length = 14)

On a failure, it will be better to return a NULL

**Comment by aballway** [comment]
>         line->_runs = runs;

The original array will be released on this assignment, but I can remove it since we are always creating it here

---
In reply to: [248395926](https://github.com/Microsoft/WinObjC/pull/989#issuecomment-248395926) [](ancestors = 248395926)

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:454 in 8c41446. [](commit_id = 8c414467615be15f6033f0d28e81ac563d40ec1d, deletion_comment = False)

File: Frameworks/CoreText/CTFrame.mm (hunk 22 above)
Comment (aballway) [inline Frameworks/CoreText/CTFrame.mm]: no, _lines is a StrongId and can't be used as a fast enumerable directly, and I prefer this to pretending it's just an NSArray*

---
In reply to: [79682448](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79682448) [](ancestors = 79682448)

**Comment by mnithish** [comment]
>         line->_runs = runs;

I missed the StrongID again - let us leave the code as is so if someone does a new will get an empty array

---
In reply to: [248397244](https://github.com/Microsoft/WinObjC/pull/989#issuecomment-248397244) [](ancestors = 248397244,248395926)

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:454 in 8c41446. [](commit_id = 8c414467615be15f6033f0d28e81ac563d40ec1d, deletion_comment = False)

File: Frameworks/CoreText/CTLine.mm [hunk 27]
```diff
@@ -367,43 +353,30 @@ CFIndex CTLineGetStringIndexForPosition(CTLineRef lineRef, CGPoint position) {
...
-        CFIndex glyphCount = CTRunGetGlyphCount(currRun);
-
-        if (currentIndex + glyphCount >= charIndex) {
-            ret = (static_cast<_CTRun*>(currRun))->_glyphOrigins[charIndex - currentIndex - 1].x;
-            if (secondaryOffset) {
-                *secondaryOffset = ret;
+    if (lineRef && charIndex >= 0) {
+        _CTLine* line = static_cast<_CTLine*>(lineRef);
+        if (charIndex > line->_strRange.location + line->_strRange.length) {
+            ret = line->_width;
```
Comment (mnithish) [inline Frameworks/CoreText/CTLine.mm]: >line->_width [](start = 18, length = 12)

Isn't the document saying it should be 0.0?

File: Frameworks/CoreText/CTRun.mm [hunk 28]
```diff
@@ -87,103 +93,89 @@ CTRunStatus CTRunGetStatus(CTRunRef run) {
...
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 const CGGlyph* CTRunGetGlyphsPtr(CTRunRef run) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    _CTRun* runPtr = static_cast<_CTRun*>(run);
+    return runPtr && runPtr->_dwriteGlyphRun.glyphCount ? runPtr->_dwriteGlyphRun.glyphIndices : nullptr;
```
Comment (mnithish) [inline Frameworks/CoreText/CTRun.mm]: >runPtr && runPtr->_dwriteGlyphRun.glyphCount [](start = 11, length = 44)

nit: put braces around this for readability

File: Frameworks/CoreText/CTFont.mm (hunk 20 above)
Comment (rajsesh-msft) [inline Frameworks/CoreText/CTFont.mm]: That is, change this to return SUCCEEDED(font->_dwriteFontFace->GetGlyphIndices(...)); instead. 

---
In reply to: [79681059](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79681059) [](ancestors = 79681059)

**Comment by rajsesh-msft** [comment]
:shipit:

File: Frameworks/CoreText/CTTypesetter.mm [hunk 29]
```diff
@@ -35,23 +35,19 @@
 
 @implementation _CTTypesetter
 - (instancetype)initWithAttributedString:(NSAttributedString*)str {
-    _attributedString = [str retain];
-    _string = [[str string] retain];
+    _attributedString.attach([str retain]);
+    _string.attach([[str string] copy]);
```
Comment (mnithish) [inline Frameworks/CoreText/CTTypesetter.mm]: nit:
_attributedString = str;
_string = [str string];

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 26 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: The reference platform will create an empty line if given "", and expects a non-null argument for CTLineCreateWithAttributedString

---
In reply to: [79685824](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79685824) [](ancestors = 79685824)

**Comment by mnithish** [comment]
> static IWLazyClassLookup s_lazyUIFont("UIFont");

I don't think we require this any more

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:40 in 8c41446. [](commit_id = 8c414467615be15f6033f0d28e81ac563d40ec1d, deletion_comment = False)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 30]
```diff
@@ -403,15 +436,16 @@ static void __GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameS
 
             // TODO::
             // This is a temp workaround until we can have actual glyph origins
-            run->_glyphOrigins.emplace_back(CGPoint{ xPos, yPos });
-            for (int index = 0; index < glyphRunDetails._dwriteGlyphRun[i].glyphCount - 1; index++) {
-                run->_glyphOrigins.emplace_back(CGPoint{ xPos + glyphRunDetails._dwriteGlyphRun[i].glyphAdvances[index], yPos });
+            for (int index = 0; index < glyphRunDetails._dwriteGlyphRun[i].glyphCount; index++) {
+                run->_glyphOrigins.emplace_back(CGPoint{ xPos, yPos });
+                run->_glyphAdvances.emplace_back(CGSize{ glyphRunDetails._dwriteGlyphRun[i].glyphAdvances[index], 0.0f });
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: >0.0f [](start = 114, length = 4)

This should be yPos right? Or does the reference platform has it all offset to 0.0?

Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: The offset is always 0.0, but so is the yPos, despite it very much not always being 0...

---
In reply to: [79692524](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79692524) [](ancestors = 79692524)

File: Frameworks/CoreText/CTLine.mm (hunk 25 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: discussed offline, reference platform returns an empty object when given an empty string

---
In reply to: [79685344](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79685344) [](ancestors = 79685344)

File: Frameworks/CoreText/CTLine.mm (hunk 27 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: discussed offline, reference platform returns line width

---
In reply to: [79688032](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79688032) [](ancestors = 79688032)

File: Frameworks/CoreText/CTFrame.mm (hunk 22 above)
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTFrame.mm]: Please DO pretend it is an NSArray.

Fast enumeration is a compiler contract optimization that allows an array to return its backing store directly, in a single message dispatch.

Non-arrays can leverage it by putting up to 16 items into a provided stack buffer, potentially reducing the message sends required to `N/16`.

An object enumerator, however, due to its design must return one object at a time. You end up using all the power of the fast enumeration for object-by-object batching.

If you want to be totally type safe and not tell any lies, cast it as such: `(id<NSFastEnumeration>)x`


---
In reply to: [79686612](https://github.com/Microsoft/WinObjC/pull/989#discussion_r79686612) [](ancestors = 79686612,79682448)

**Comment by DHowett-MSFT** [comment]
:clock1:
